[Form] [Extension] [FormTypeCsrfExtension] Make the csrf field options extensible

esserj opened the pull request:
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR | none |

If you wana decorate the FormTypeCsrfExtension you have no control over options set by the extension, forcing you to reimplement the entire finishView. this way you have a bit more control over it. We required this approach to define custom options declared by another extension

esserj committed:
Make the csrf field options extensible

if you wana decorate the FormTypeCsrfExtension you have no control over options set by the extension, forcing you to reimplement the entire finishView. this way you have a bit more control over it. We required this approach to define custom options declared by another extension

esserj committed:
replaces new by old array notation

esserj commented:
fixed the fail from fabbot.io

jakzal commented:
Just so I could understand the use case better, what kind of option you needed to customize?

esserj commented:
We build an extension that requires a by the extension defined option to be set on all fields in order to activate itself on that field. its a client side UI handling submissions through ajax, etc.. we want this process to be opt-in thats why doing it for every formtype by default is not an option

ro0NL commented on src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php:
```diff
@@ -124,6 +124,7 @@ public function configureOptions(OptionsResolver $resolver)
             'csrf_message' => 'The CSRF token is invalid. Please try to resubmit the form.',
             'csrf_token_manager' => $this->defaultTokenManager,
             'csrf_token_id' => null,
+            'csrf_field_options' => array()
```
Perhaps define this below `csrf_field_name` for clarification.

ro0NL commented on src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php:
```diff
@@ -105,9 +105,9 @@ public function finishView(FormView $view, FormInterface $form, array $options)
             $data = (string) $options['csrf_token_manager']->getToken($tokenId);
 
-            $csrfForm = $factory->createNamed($options['csrf_field_name'], 'Symfony\Component\Form\Extension\Core\Type\HiddenType', $data, array(
+            $csrfForm = $factory->createNamed($options['csrf_field_name'], 'Symfony\Component\Form\Extension\Core\Type\HiddenType', $data, array_replace($options['csrf_field_options'], array(
```
Imo. `array_replace_recursive` would be considerable here.

ro0NL replied:
Also i'd toggle the arguments, so we can actually override the `mapped` option if needed.

ro0NL commented:
Perhaps adding `csrf_field_type` makes sense as well.. so we are fully flexible.

esserj committed:
Changed order of arguments and added support for custom form type

csrf_field_type added, csrf_field_options moved to where related options are and it is now possible to also override the mapped option

esserj commented:
I've added the suggestions, I see where you wana go with array_replace_recursive, but it seems only usefulll if you by default have a multi-dimentional array and mapped is not, i'd keep that for when you make changes to the default array that require recursive replacements?

ro0NL commented on src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php:
```diff
@@ -121,9 +121,11 @@ public function configureOptions(OptionsResolver $resolver)
             'csrf_message' => 'The CSRF token is invalid. Please try to resubmit the form.',
             'csrf_token_manager' => $this->defaultTokenManager,
-            'csrf_token_id' => null,
+            'csrf_token_id' => null
```
Last comma is required code style, you should revert this line :)

esserj committed:
reverted removal of code-style required trailing array comma

esserj commented:
fixed that, must take some time to go through the code style, only jumped on the symfony wagon recently :)

patrick-mcdougle commented:
@esserj 

>we want this process to be opt-in thats why doing it for every formtype by default is not an option

Is this different than disabling it?

esserj commented:
@patrick-mcdougle an opt-out strategy would in our case have the same issue

patrick-mcdougle commented:
I mean, how is this different than disabling csrf on some of your forms? Or am I not understanding the use case?

esserj commented:
Why would we want to disable CSRF? This is a CMS back-end where admins will be working in. If there is one thing we want in there its CSRF. so we would have to roll my own version or decorate the current implementation (which we did now) but any future changes/improvements will have to manually be ported. And thats when I decided to create the PR as it will address the non-extensible nature of the current implementation

patrick-mcdougle commented:
I'm just not clear on what you're trying to do. I didn't follow your explanation to @jakzal Why do you need the csrf token field to be rendered using something different than a hidden field?

esserj commented:
@patrick-mcdougle I don't need anything different than a hidden field, that was a suggestion from @ro0NL that i took up in this extensibility patch, i need to be able to add an option to the csrf field

xabbuh commented on src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php:
```diff
@@ -105,9 +105,9 @@ public function finishView(FormView $view, FormInterface $form, array $options)
             $data = (string) $options['csrf_token_manager']->getToken($tokenId);
 
-            $csrfForm = $factory->createNamed($options['csrf_field_name'], 'Symfony\Component\Form\Extension\Core\Type\HiddenType', $data, array(
+            $csrfForm = $factory->createNamed($options['csrf_field_name'], $options['csrf_field_type'], $data, array_replace(array(
```
While I can understand that there may be use case where you need to (want to) customise the CSRF field's options, I do not see a need to change the type being used. Without a really good use case I am against making it configurable.

ro0NL replied:
Simple, allow for `MyHiddenType::class`. Imo. it's weird to allow passing options, but not allowing a custom type with _default_ options.

xabbuh replied:
There shouldn't be a custom hidden type. If you are creating a custom hidden type, you will very likely better be off with a form type extension.

ro0NL replied:
Maybe your're right.. i have no actual case for this. Then again, forcing a specific type doesnt feel extensible for an _extension_...

fabpot replied:
I agree with @xabbuh

HeahDude commented:
Hi @esserj,  thank you for proposing this change, but the use case it supported can already be solved.

>If you wana decorate the FormTypeCsrfExtension you have no control over options set by the extension,

Since 3.2 it is possible to handle priorities on form type extension when registering them as services, this feature has been added so it can be easier to rely on other extensions definitions.

Hence I don't think we need to add yet another option to handle this, you could easily override options from another type extension now. Using the Form component stand alone should already provide full control on the registration order of different type extensions.

If there is no other concrete use cases, I'm 👎 for now.

fabpot commented:
Closing this PR as there is no more activity and several concerns were raised.
